### ISSUE TITLE
Docker API does not report correct images for /containers/json endpoint

### ISSUE THREAD

**Issue by Luzifer** [docker/docker#12595]
# Description of problem

Docker API (`v.1.18`) does not report correct images for `/containers/json` endpoint. Instead of the actual image ID for an image not longer tagged the previous tag is reported.

# `docker version`

```
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64
```

# `docker info`

```
Containers: 6
Images: 139
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 151
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-49-generic
Operating System: Ubuntu 14.04.2 LTS
CPUs: 8
Total Memory: 31.29 GiB
Name: Ubuntu-1404-trusty-64-minimal
ID: QE6C:ZNQX:BADL:4BTY:7AK5:274U:7B7N:EUS7:FGYI:TE5Q:JPXI:FBK6
WARNING: No swap limit support
```

# `uname -a`

```
Linux Ubuntu-1404-trusty-64-minimal 3.13.0-49-generic #83-Ubuntu SMP Fri Apr 10 20:11:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

# Environment details (AWS, VirtualBox, physical, etc.):

Bare-Metal, Ubuntu, latest lxc-docker

# How reproducible

100%

# Steps to Reproduce:
1. `docker pull` an image
2. `docker run` it (and give the container a name)
3. Update the image
4. `docker pull` the changed image
5. Ask the `/containers/json` endpoint about the container

# Actual Results

- The image ID for the tag has changed
- The old image is shown with "Repository" `<none>` 
- The API reports the old tag for the image the container is running with

# Expected Results

- The image ID for the tag has changed
- The old image is shown with "Repository" `<none>` 
- The API reports the **image ID** for the image the container is running with

# Additional info

```
# curl 127.0.0.1:21/containers/json?all=1
[{"Command":"./gobuilder","Created":1429599867,"Id":"dc520983d708749528b900118383ac20143a657a156c5efadc678892d24215e3","Image":"luzifer/gobuilder-frontend:latest","Labels":{},"Names":["/gobuild-frontend"],"Ports":[{"IP":"0.0.0.0","PrivatePort":3000,"PublicPort":1060,"Type":"tcp"}],"Status":"Up 5 hours"}]
# docker inspect gobuild-frontend | grep -i image
        "Image": "luzifer/gobuilder-frontend:latest",
    "Image": "b8e06e06802772cad27d271eac378eca042ebac6e3653bea21bd0751e1f30dfa",
# docker images
REPOSITORY                   TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
luzifer/gobuilder-frontend   latest              b96f5f028d3c        2 hours ago         592 MB
<none>                       <none>              b8e06e068027        5 days ago          591.7 MB
```

**Comment by Luzifer** [comment]
Until Docker 1.5.0 the expected behavior was present

**Comment by jperville** [comment]
+1, this bug prevents our chef cookbooks from detecting that a container should be redeployed after the image has been updated.

**Comment by abrisse** [comment]
+1 very annoying regression

**Comment by jperville** [comment]
Here is a short bash example demonstrating the issue.

If I update the image of a docker container (eg. by doing "docker pull" after the container has run) the container still reports that it follows the tagged image name, while it should be displaying the internal image identifier (since the tag has moved).

```sh
# container=myapp-1
# container_image_id=`docker inspect --format '{{ .Image }}' ${container}`
# image_id=`docker inspect --format '{{ .Id }}' $(docker ps | grep ${container} | awk '{print $2}')`
# echo "Must be EQUAL ${container_image_id} AND ${image_id}"
Must be EQUAL 608e6e154266bdc6a7429f969233ea2378ac43fb88868adbbcdef1b87cb3167c AND 608e6e154266bdc6a7429f969233ea2378ac43fb88868adbbcdef1b87cb3167c
# docker tag -f 81e3cc14a804 myrepo/myapp
# container_image_id=`docker inspect --format '{{ .Image }}' ${container}`
# image_id=`docker inspect --format '{{ .Id }}' $(docker ps | grep ${container} | awk '{print $2}')`
# echo "Must be EQUAL ${container_image_id} AND ${image_id}"
Must be EQUAL 608e6e154266bdc6a7429f969233ea2378ac43fb88868adbbcdef1b87cb3167c AND 81e3cc14a804a7e424201af69faf41b39f70dec18aa9ea19334b7f3662f98601
```

The consequence is that my cookbook should rely on the real image ID (reported by `docker inspect --format '{{ .Image }}' ${container}`) and not on whatever image is listed in the output of `docker ps`.

**Comment by cpuguy83** [comment]
Confirmed, this is because `ps` is using container.Config.Image instead of container.Image to ensure we get the correct tag name... unfortunately tags can be changed.

**Comment by cpuguy83** [comment]
This should check if the ID for container.Config.Image matches container.Image before using it.

**Comment by relistan** [comment]
This looks like it was broken in #11895, with the specific change [here](11895)

**Comment by cpuguy83** [comment]
@relistan issue was that "docker ps" was reporting a different name than was used to create the container sometimes when there are more than 1 tag on the image.
Now it is just showing whatever was passed in when creating the image, but this gets out of date.

**Comment by LK4D4** [comment]
@relistan Original issue was IIRC that it tried to find `ImageName` from `ImageID`, but there is can be multiple tags with same id and you can see in ps that you created container with another tag, tag that you created.

**Comment by LK4D4** [comment]
Now it just shows what you write in docker run command.

**Comment by didip** [comment]
For completion sake, this bug also occur on `1.5.0-dev` version.
**docker version**
```
Client version: 1.5.0-dev
Client API version: 1.18
Go version (client): go1.3.3
Git commit (client): fc0329b/1.5.0
OS/Arch (client): linux/amd64
Server version: 1.5.0-dev
Server API version: 1.18
Go version (server): go1.3.3
Git commit (server): fc0329b/1.5.0
OS/Arch (server): linux/amd64
```

**docker ps**
```
docker ps
CONTAINER ID        IMAGE                                                                     COMMAND                CREATED             STATUS              PORTS                                                                                            NAMES
4eb036e6a545        31cc81ffb6356211de6e969292abc7b5620a9cab137e46565ddc97e60412aef3:latest   "supervisord"          About an hour ago   Up 10 minutes       10.1.23.24:10200-10208->10200-10208/tcp, 10.1.23.24:10211->10211/tcp                             xxxxx-service-7bd4107333e13e                  
```

**Comment by relistan** [comment]
@cpuguy83 Thanks! I see. :)

@LK4D4 That's really undesirable behavior because tags/image names move. It's unexpected for something to tell me it's tagged "production" when it's not running the current "production". Or "latest" when it's not the one tagged "latest". 

Here's a Docker 1.5 install showing a container for which the image name and tag have moved:
```bash
$ docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS                                   NAMES
8d826d1245c4        2485b0f89951        "nginx -g 'daemon of   9 weeks ago         Up 2 seconds        443/tcp, 172.16.168.179:10234->80/tcp   serene_banach
```

Here's what that looks like on 1.6:
```bash
$ docker ps
CONTAINER ID        IMAGE                                                                     COMMAND                CREATED             STATUS              PORTS                                   NAMES
8d826d1245c4        2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d:latest   "nginx -g 'daemon of   9 weeks ago         Up 3 seconds        443/tcp, 172.16.168.179:10234->80/tcp   serene_banach
```
Notice how it's doing a few wrong things. First, it's showing the whole long image ID. Second, it's showing that it has the `latest` tag which it does not.

Here's what the images look like on that same host:
```bash
$ docker images
REPOSITORY    TAG                 IMAGE ID            CREATED              VIRTUAL SIZE
nginx         latest              224873bdcaa1        4 weeks ago          93.44 MB
<none>        <none>              2485b0f89951        10 weeks ago         93.46 MB
```
Note that the `latest` tag is supposed to be on *224873bdcaa1* not on *2485b0f89951*.

It's moderately annoying to have a different tag show up than the one I invoked it with. But it's actually a production issue for us if the tag is showing incorrectly.

**Comment by thieman** [comment]
I'm trying to work on this. New to the project, so didn't want to call dibs in case someone better wanted to take a look.

I think I've got a failing test case together if anyone wants to take a look at it. https://gist.github.com/thieman/331edd7e2a3c6b397663

**Comment by relistan** [comment]
I've got a first pass at code that seems to work. Testing some more. @thieman giving your test a whirl now.

**Comment by relistan** [comment]
Oh I see it looks like you committed a fix already.

**Comment by thieman** [comment]
Been hitting a bunch of edge cases with `LookupImage`. Think this latest test run might work, though. Care to share what you've got?

**Comment by thieman** [comment]
The commit linked up there is wrong, btw, it should properly be linking to this entire PR: https://github.com/docker/docker/pull/12918

**Comment by relistan** [comment]
I've moved my commentary to that PR.

**Comment by relistan** [comment]
Working from @thieman's test case on #12918 I think there are two different problems here:

1. I update an image tag following the steps in @thieman's test case. I end up with a container with the wrong ImageID:

  ```
  $ docker tag busybox:latest busybox:TestPsImageIDAfterUpdate-original
  $ docker run -d -t busybox:TestPsImageIDAfterUpdate-original /bin/sh
  $ docker commit 40c4ffc128c8 busybox:TestPsImageIDAfterUpdate-updated
  $ docker inspect 40c4ffc128c8 | grep -i image
          "Image": "busybox:TestPsImageIDAfterUpdate-original",
      "Image": "8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55",
  $ docker images | grep busy
  busybox                              TestPsImageIDAfterUpdate-updated    463f1c28e186        18 minutes ago      2.433 MB
  busybox                              TestPsImageIDAfterUpdate-original   8c2e06607696        13 days ago        2.433 MB
  busybox                              latest                              8c2e06607696        13 days ago         2.433 MB
```
That's bad because the Image *should* be *463f1c28e186* not *8c2e06607696*.  I believe this is a problem with the code that commits a container and is not updating the ImageID appropriately

2. The issue that @didip and I were reporting (which I _think_ is the same as @luzifer's issue) which is that when an image name/tag move due to a `docker pull`, you end up with the tag remaining attached rather than the bare ID like this:

    1. Image exists as `busybox`
    2. I start up a container
    3. docker ps shows my image as `busybox:latest`
    4. `busybox `is outdated and I do a docker pull which updates the image
    5. `docker ps` shows my image ID as *2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d:latest*

That image *should* be *2485b0f8995167b796a7ab4696e842c08bd27d58febb3ba0e935a4e2c1b2807d* . This I believe is a display issue with the code I linked earlier.

**Comment by thieman** [comment]
This makes sense to me, I interpreted the problem to be your Issue 1, which is why I've been acting like a confused idiot as we've been dancing around Issue 2. :smile: 

Is my PR an appropriate fix for your Issue 1? If so, should we try to get that merged, then either create a new issue or clean this one up to be explicitly about Issue 2? Having the two simultaneously being discussed in #12595 is boggling my mind.

**Comment by relistan** [comment]
@thieman, going to validate your code with our issue now.

**Comment by relistan** [comment]
Confirmed, @thieman's PR #12918 solves issue (2) above. *but* this code, like the code committed in #11895 changes the behavior vs 1.5.0 of images that do not have a tag. The old behavior would append *:latest* and that is no longer the case. @LK4D4 was the change to displaying long image ID's intentional, or is that also a bug?

@thieman I'll try to work up an integration test to validate (2) and submit a PR on yours

**Comment by thieman** [comment]
@relistan A note on your comment illustrating the two issues: your steps in issue 1 are not reflecting my test case. You need to overwrite the `busybox:TestPsImageIDAfterUpdate-original` image, but in your third line you are creating the `-updated` image. My step uses a `tag -f` to do this after using the commit to create the `-updated` image.

Your issue is showing that the image value in `ps` remains the original tag. I think this is probably fine since that original tag is still valid, and that is after all the tag with which the image is created. That's more of a design question, though.

**Comment by thieman** [comment]
I know I just said this was a design decision, but based on a reading of the `commit` documentation I do not think it makes sense for the running container's `ImageID` to be changed after a commit. The `--change` option indicates that it applies Dockerfile commands to the *image* that is created. These commands would cause your new image to diverge from what is actually running in your current container. As such, it would be false for the running container to then indicate that it was based off of your new image.

https://docs.docker.com/reference/commandline/cli/#commit

**Comment by dteeter** [comment]
I am seeing this problem with docker version 1.8.1 and api version 1.20.

A first container listing after creating the container had the correct image name and tag in the "Image" property. However, in a another listing less than a second later, the "Image" property had changed to the image id and continued to be the image id in subsequent listings. For me the only workaround is to use labels to keep track of the container's image name and tag.

**Comment by dteeter** [comment]
It seems that either I do not have rights to reopen this bug or it cannot be reopened. Should I create a new one?

**Comment by cpuguy83** [comment]
@dteeter the referenced PR that closed this issue was included in docker 1.9

Please do not create a new bug, it is already fixed.

**Comment by dteeter** [comment]
OK, thanks for the info
